// @octoai/client/build/module/lib/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const version_1 = require("../version");
const asset_1 = require("./asset");
const chat_1 = require("./chat");
const completions_1 = require("./completions");
const error_1 = require("./error");
const tune_1 = require("./tune");
/**
 * A client that allows inferences from existing OctoAI endpoints.  Sets
 * various headers, establishes clients for {@link Chat} under `Client.chat`,
 * {@link AssetLibrary} under `Client.asset`, {@link FineTuningClient} under
 * `Client.tune`, and will check for `OCTOAI_TOKEN`
 * from environment variable if no token is provided.
 *
 * @throws {@link OctoAIClientError} - For client-side failures (throttled, no token)
 * @throws {@link OctoAIServerError} - For server-side failures (unreachable, etc)
 *
 * @remarks
 * You can create an OctoAI API token by following the guide at
 * {@link https://docs.octoai.cloud/docs/how-to-create-an-octoai-access-token |
 * How to Create an OctoAI Access Token}
 */
class Client {
    /**
     * Constructor for the Client class.
     *
     * @param token - OctoAI token.  If none is set, checks for an `OCTOAI_TOKEN`
     * envvar, or will default to null.
     * @param secureLink - Set to true to use SecureLink API instead of public API
     */
    constructor(token, secureLink = false) {
        token = token || process.env.OCTOAI_TOKEN || null;
        if (!token) {
            console.warn("OCTOAI_TOKEN is not set as an environment variable and " +
                "wasn't passed to the Client constructor as a token." +
                " You will only be able to use QuickStart OctoAI endpoints.  " +
                "client.asset, client.chat, and client.tune will throw errors if " +
                "accessed.");
        }
        const headers = {
            "Content-Type": "application/json",
            "User-Agent": `octoai-ts-${version_1.LIB_VERSION}`,
            Authorization: "",
            "X-OctoAI-Async": "", // 0 disabled, 1 enabled
            Accept: "",
        };
        if (token) {
            headers.Authorization = `Bearer ${token}`;
        }
        this.secureLink = secureLink;
        this.headers = headers;
        this.chat = new chat_1.Chat(this);
        this.asset = new asset_1.AssetLibrary(this);
        this.tune = new tune_1.FineTuningClient(this);
        this.completions = new completions_1.CompletionsAPI(this);
    }
    /**
     * Send a request to the given endpoint with inputs as request body.
     * For LLaMA2 LLMs, this requires `"stream": false` in the inputs.  To stream
     * for LLMs, please see the {@link inferStream} method.
     *
     * @param endpointUrl - Target URL to run inference
     * @param inputs - Necessary inputs for the endpointURL to run inference
     *
     * @returns JSON outputs from the endpoint
     */
    async infer(endpointUrl, inputs) {
        const response = await fetch(endpointUrl, {
            method: "POST",
            headers: this.headers,
            body: JSON.stringify(inputs),
        });
        if (!response.ok) {
            await (0, error_1.throwOctoAIError)(response);
        }
        return response.json();
    }
    /**
     * Stream text event response body for supporting endpoints.  This is an
     * alternative to loading all response body into memory at once.  Recommended
     * for use with LLM models.  Requires `"stream": true` in the inputs for
     * LLaMA2 LLMs.
     *
     * @param endpointUrl - Target URL to run inference
     * @param inputs - Necessary inputs for the endpointURL to run inference
     * @returns Compatible with getReader method.
     *
     * @remarks
     * This allows you to stream back tokens from the LLMs.  Below is an example
     * on how to do this with a LLaMA2 LLM using a completions style API.
     *
     * HuggingFace style APIs will usually use the variable `done` below to
     * indicate the end of the stream.  OpenAI style APIs will often send a
     * string in the stream `"data: [DONE]\n"` to indicate the stream is complete.
     *
     * This example concatenates all values from the tokens into a single text
     * variable.  How you choose to use the tokens will likely be different, so
     * please modify the code.
     *
     * This examples assumes:
     * 1) You've followed the guide at
     * {@link https://docs.octoai.cloud/docs/how-to-create-an-octoai-access-token |
     * How to Create an OctoAI Access Token} to create and set your OctoAI access
     * token
     * 2) Either that you will set this token as an OCTOAI_TOKEN envvar
     * or edit the snippet to pass it as a value in the `{@link Client.constructor}`.
     * 3) You have assigned your endpoint URL and inputs into variables named
     * llamaEndpoint and streamInputs.
     *
     *```ts
     * const client = new Client();
     *     const readableStream = await client.inferStream(
     *       llamaEndpoint,
     *       streamInputs
     *     );
     * let text = ``;
     * const streamReader = readableStream.getReader();
     * for (
     *   let { value, done } = await streamReader.read();
     *   !done;
     *   { value, done } = await streamReader.read()
     * ) {
     *   if (done) break;
     *   const decoded = new TextDecoder().decode(value);
     *   if (
     *     decoded === "data: [DONE]\n" ||
     *     decoded.includes('"finish_reason": "')
     *   ) {
     *     break;
     *   }
     *   const token = JSON.parse(decoded.substring(5));
     *   if (token.object === "chat.completion.chunk") {
     *     text += token.choices[0].delta.content;
     *   }
     * console.log(text);
     *```
     * The `const token = JSON.parse(decoded.substring(5))` line strips `"data"`
     * from the returned text/event-stream then parses the token as an object.
     */
    async inferStream(endpointUrl, inputs) {
        const headers = { ...this.headers, Accept: "text/event-stream" };
        const response = await fetch(endpointUrl, {
            method: "POST",
            headers: headers,
            body: JSON.stringify(inputs),
        });
        if (!response.ok) {
            await (0, error_1.throwOctoAIError)(response);
        }
        return response;
    }
    coldStartWarning() {
        console.warn("Your endpoint may take several minutes to start " +
            "and be ready to serve inferences. You can increase your endpoint's " +
            "min replicas to mitigate cold starts.");
    }
    /**
     * Check health of an endpoint using a get request.  Try until timeout.
     *
     * @param endpointUrl - Target URL to run the health check.
     * @param timeoutMS - Milliseconds before request times out.  Default is 15
     * minutes.
     * @param intervalMS - Interval in milliseconds before the healthCheck method
     * queries
     * @returns HTTP status code.
     *
     * @remarks
     * The default timeout is set to 15 minutes to allow for potential cold start.
     *
     * For custom containers, please follow
     * {@link https://docs.octoai.cloud/docs/health-check-paths-in-custom-containers
     * | Health Check Paths in Custom Containers} to set a health check endpoint.
     *
     * Information about health check endpoint URLs are available on relevant
     * QuickStart Templates.
     */
    async healthCheck(endpointUrl, timeoutMS = 900000, // 15 minutes for cold start
    intervalMS = 1000) {
        const start = new Date().getTime();
        // Fence post
        let response = await fetch(endpointUrl, {
            method: "GET",
            headers: this.headers,
        });
        if (!response.ok) {
            this.coldStartWarning();
        }
        while (new Date().getTime() - start < timeoutMS && !response.ok) {
            if (400 <= response.status && response.status < 500) {
                await (0, error_1.throwOctoAIError)(response);
            }
            response = await fetch(endpointUrl, {
                method: "GET",
                headers: this.headers,
            });
            await new Promise((resolve) => setTimeout(resolve, intervalMS));
        }
        if (response.status >= 400) {
            await (0, error_1.throwOctoAIError)(response);
        }
        return response.status;
    }
    /**
     * Execute an inference in the background on the server.
     *
     * @param endpointUrl - Target URL to send inference request.
     * @param inputs - Contains necessary inputs for endpoint to run inference.
     * @returns Future allows checking if results are ready then accessing them.
     *
     * @remarks
     * Please read the {@link https://docs.octoai.cloud/reference/inference |
     * Async Inference Reference} for more information.
     * {@link Client.inferAsync} returns an {@link InferenceFuture},
     * which can then be used with {@link Client.isFutureReady} to see the
     * status.  Once it returns `true`, you can use the
     * {@link Client.getFutureResult} to get the response for your
     * InferenceFuture.
     *
     * Assuming you have a variable with your target endpoint URL and the inputs
     * the model needs, and an `OCTOAI_TOKEN` set as an environment variable, you
     * can run a server-side asynchronous inference from
     * {@link https://docs.octoai.cloud/docs/welcome-to-the-octoai-compute-service-copy | QuickStart Template}
     * endpoints with something like the below.
     *
     * ```ts
     *  const client = new Client();
     *  const future = await client.inferAsync(url, inputs);
     *  if (await client.isFutureReady(future) === true) {
     *    return await client.getFutureResult(future);
     *  }
     * ```
     */
    async inferAsync(endpointUrl, inputs) {
        const headers = { ...this.headers };
        headers["X-OctoAI-Async"] = "1"; // Enables server-side async
        const response = await fetch(endpointUrl, {
            method: "POST",
            headers: headers,
            body: JSON.stringify(inputs),
        });
        if (response.status >= 400) {
            await (0, error_1.throwOctoAIError)(response);
        }
        const future = (await response.json());
        return future;
    }
    async pollFuture(future) {
        const response = await fetch(future.poll_url, {
            method: "GET",
            headers: this.headers,
        });
        if (response.status >= 400) {
            await (0, error_1.throwOctoAIError)(response);
        }
        return (await response.json());
    }
    /**
     * Return whether the {@link InferenceFuture} generated from
     * {@link Client.inferAsync} has been computed and can return results.
     *
     * @param future - Created from {@link Client.inferAsync}.
     * @returns True if the {@link InferenceFuture}
     * inference is completed and are able to use {@link Client.getFutureResult}.
     * Else returns false.
     */
    async isFutureReady(future) {
        const respObject = await this.pollFuture(future);
        return "completed" === respObject.status;
    }
    /**
     * Return the result of a {@link InferenceFuture} generated from
     * {@link Client.inferAsync} as long as {@link Client.isFutureReady} returned
     * `true`.
     *
     * @param future - An {@link InferenceFuture} generated from
     * {@link Client.inferAsync}
     * @returns JSON outputs from the endpoint.
     */
    async getFutureResult(future) {
        const respObject = await this.pollFuture(future);
        if (respObject.status !== "completed") {
            await (0, error_1.throwOctoAIError)(new Response(null, {
                status: 425,
                statusText: "Wait until isFutureReady returns true before running getFutureResult.",
            }));
        }
        const responseUrl = respObject.response_url;
        const response = await fetch(responseUrl, {
            method: "GET",
            headers: this.headers,
        });
        if (response.status >= 400) {
            await (0, error_1.throwOctoAIError)(response);
        }
        return response.json();
    }
}
exports.Client = Client;
