// @octoai/client/build/module/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwOctoAITokenRequiredError = exports.throwOctoAIError = exports.OctoAITimeoutError = exports.OctoAIServerError = exports.OctoAIClientError = exports.OctoAIValidationError = void 0;
/**
 * A class of errors specific to the OctoAI TS SDK.
 */
class OctoAIError extends Error {
    /**
     * Parent class for OctoAI Errors used by the TS SDK.
     *
     * @param message - Message from server or client about cause of error.
     * @param causedBy - Error code of course error.  Usually HTTP errors.
     */
    constructor(message, causedBy) {
        super(message);
        this.name = this.constructor.name;
        this.causedBy = causedBy;
    }
}
/**
 * Client-side validation errors.  Problems with request caught before sending.
 */
class OctoAIValidationError extends OctoAIError {
}
exports.OctoAIValidationError = OctoAIValidationError;
/**
 * Client-side errors.  Indicated by HTTP codes from 400 through 499.
 */
class OctoAIClientError extends OctoAIError {
}
exports.OctoAIClientError = OctoAIClientError;
/**
 * Server-side errors.  Indicated by HTTP codes from 500 through 599.
 */
class OctoAIServerError extends OctoAIError {
}
exports.OctoAIServerError = OctoAIServerError;
/**
 * Timeout errors that don't necessarily indicate a client or server failure.
 */
class OctoAITimeoutError extends OctoAIError {
}
exports.OctoAITimeoutError = OctoAITimeoutError;
/**
 * Support method to create errors to throw on failures.
 *
 * @param response - the server response that isn't ok.
 */
async function throwOctoAIError(response) {
    const statusCode = response.status;
    let text = response.statusText;
    try {
        text += ": " + JSON.stringify(await response?.json());
    }
    catch (e) {
        // Response doesn't have JSON
    }
    let reqIDStr = "";
    const xUserRequestID = response.headers?.get("X-User-Request-ID");
    if (xUserRequestID !== null && xUserRequestID !== undefined) {
        reqIDStr = ` (request_id: ${response.headers.get("X-User-Request-ID")})`;
    }
    text += reqIDStr;
    if (statusCode >= 500) {
        throw new OctoAIServerError(`Server error: ${statusCode} ${text}`);
    }
    else if (statusCode === 429) {
        throw new OctoAIClientError(`Rate limit reached error: ${statusCode} ${text}`);
    }
    else {
        throw new OctoAIClientError(`Error: ${statusCode} ${text}`);
    }
}
exports.throwOctoAIError = throwOctoAIError;
/**
 * Standardized error message for requiring authorization.
 */
function throwOctoAITokenRequiredError() {
    const msg = "An OctoAI token is required to use this API.  Please " +
        "pass your OctoAI token to the client class during instantiation or set it " +
        "as an environment variable named `OCTOAI_TOKEN`.";
    throw new OctoAIValidationError(msg);
}
exports.throwOctoAITokenRequiredError = throwOctoAITokenRequiredError;
